### PR TITLE
[DCP fixes for _load_state_dict_keys and supports nested keys

### PR DESCRIPTION
**Pull request by LucasLLC**
Fixes some issues with `_load_state_dict_keys`, including:
  * updates broken test, which was failing due to incorrect parameters
  * adds support for specifying nested keys e.g. (load state dict keys can now specify `something like "optimizer.state"`, which loads all keys under `optimzier.state`.
  * updates call site to use the private implementation of `_load_state_dict`, which properly handles empty state dicts (otherwise the keys are ignored)

Big shout out to @diego-urgell who not only identified current issues, but recommended the right solutions!

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Comment by fegin** [comment]
Do you mind to add description of what is the bug?

File: torch/distributed/checkpoint/default_planner.py [hunk 1]
```diff
@@ -234,6 +234,29 @@ def __init__(self, keys=None, *args, **kwargs):
...
+        for unflattened_key in planner_data:
+            if unflattened_keys:
+                unflattened_keys.append(
+                    ".".join([unflattened_keys[-1], unflattened_key])
+                )
+
+            else:
+                unflattened_keys.append(unflattened_key)
+
+        if any(unflattened_key in self.keys for unflattened_key in unflattened_keys):
```
Comment (diego-urgell) [inline torch/distributed/checkpoint/default_planner.py:255]: 🔥

File: torch/distributed/checkpoint/default_planner.py [hunk 2]
```diff
@@ -234,6 +234,29 @@ def __init__(self, keys=None, *args, **kwargs):
...
         super().__init__(*args, **kwargs)
 
+    def _should_include_key(self, key: str, metadata: Metadata) -> bool:
+        if self.keys is None:
+            return True
+
+        if key in self.keys:
+            True
+
+        unflattened_keys = []
```
Comment (diego-urgell) [inline torch/distributed/checkpoint/default_planner.py:244]: Nit: I wonder if something like "unflattened_intermediary_paths" could be a more intuitive name for this variable?

**Review by diego-urgell — APPROVED** [review]
Thank you for implementing this so fast! 😄

File: torch/distributed/checkpoint/state_dict_loader.py [hunk 3]
```diff
@@ -283,12 +285,26 @@ def _load_state_dict_from_keys(
         "torch.distributed.checkpoint._load_state_dict_from_keys"
     )
 
+    no_dist = not (dist.is_available() and dist.is_initialized())
+    if no_dist:
+        warnings.warn(
+            "torch.distributed is unavailable or uninitialized, assuming the intent is to load in a single process."
```
Comment (Skylion007) [inline torch/distributed/checkpoint/state_dict_loader.py:291]: Mind setting a stack level for this warning?

**Review by wz337 — APPROVED** [review]
LGTM!

File: torch/distributed/checkpoint/state_dict_loader.py (hunk 3 above)
Comment (LucasLLC) [reply torch/distributed/checkpoint/state_dict_loader.py:291]: Hey @Skylion007 thanks for the suggestion, I'll see if I can circle back to this in a follow-up and update all of our warnings.

**Pull request closed without merging**